src/main/java/org/folio/rest/impl/LoansAPI.java

```diff
@@ -293,9 +294,30 @@ private Errors moreThanOneOpenLoanError(Loan entity) {
   }
 
   private boolean isMultipleOpenLoanError(AsyncResult<Response> reply) {
-    return reply.succeeded()
-      && reply.result().getStatus() == 400
-      && reply.result().getEntity().toString().contains("loan_itemid_idx_unique");
+    String message = "";
+
+    if (reply.succeeded() && reply.result().getStatus() >= 400 &&
+      reply.result().getStatus() < 600 && reply.result().hasEntity()) {
+
+      // When entity is an instance of Errors, the getEntity().toString()
+      // will return an object identifier. Process the object to correctly
+      // parse the message.
+      if (reply.result().getEntity() instanceof Errors) {
+        Errors errors = (Errors) reply.result().getEntity();
+
+        for (int i = 0; i < errors.getErrors().size(); i++) {
+          Error error = errors.getErrors().get(i);
+
+          if (error.getType() == RTFConsts.VALIDATION_FIELD_ERROR)
+            message = error.getMessage().toLowerCase();
+        }
+      } else {
+        message = reply.result().getEntity().toString().toLowerCase();
+      }
+    }
+
+    return message.contains("value already exists in table loan: ")
+      || message.contains("duplicate key value violates unique constraint ");
```

Comment:
My interpretation of this change is that the description of the error is more stable than the name of the index. Is that a reasonable understanding?

What happens if the loan table has more than one unique constraint, will this code think that it is multiple open loans?

Reply:
Yes/No, the index is no longer appearing in the error message in some circumstances.

This messsage itself is (sometimes) also being added: https://github.com/folio-org/raml-module-builder/blob/master/domain-models-runtime/src/main/java/org/folio/rest/persist/PgUtil.java#L332 .

It could be that this should be now using `table:value` syntax instead of index in addition to the generic error message string `value already exists in table loan:`.
The problem I see then is that we need to change how that error is caught in the first place.
If I remember correctly, there are different tables/values/indexes that would get called by say `isMultipleOpenLoanError`.

It might indeed make more sense to write this as a utility (as requested) and then require a seperate parameter for the specific _table_ and the specific _value_ to be matched.

Reply:
> Yes/No, the index is no longer appearing in the error message in some circumstances.
> 
> This messsage itself is (sometimes) also being added: https://github.com/folio-org/raml-module-builder/blob/master/domain-models-runtime/src/main/java/org/folio/rest/persist/PgUtil.java#L332 .

Hmm, it is interesting that the error message is different even under the same circumstances. Can it be confirmed that the underlying cause is the same unique constraint in all cases?

> It could be that this should be now using `table:value` syntax instead of index in addition to the generic error message string `value already exists in table loan:`.
> The problem I see then is that we need to change how that error is caught in the first place.

Can you expand upon what you mean here? Do you mean the error should be reported differently from PgUtil?

> If I remember correctly, there are different tables/values/indexes that would get called by say `isMultipleOpenLoanError`.

The intent of `isMultipleOpenLoanError` was to infer from an opaque error message whether that was the underlying cause of the error.

I think we still need a way to achieve that. I don't know what information we have available in order to make that decision.

> It might indeed make more sense to write this as a utility (as requested) and then require a seperate parameter for the specific _table_ and the specific _value_ to be matched.

If we can extract the general parts of this into a common function with some parameters that would be great.

Some work to do that was done in a special matcher for tests, which might be worth finding. It wouldn't be appropriate to use, as it is test specific.

Reply:
> Can you expand upon what you mean here? Do you mean the error should be reported differently from PgUtil?

I was thinking more along the lines of changing the tests to look specifically for their table:value (aka: leave PgUtil alone).
Now that I am re-looking at the error message `value already exists in table loan: 335b7118-76e7-4d0a-9453-3dda29dfe529` is not that much more helpful. 

The error message `...value already exists in table loan:...` (HTTP 422) does not contain the unique index, and so there is no way to truly infer this.

The error message `...duplicate key value violates unique constraint "loan_itemid_idx_unique"...` (HTTP 400) does contain the index and can be inferred.

The other error message seems to be with `LoansPolicyApi.java` and not `LoansApi.java`.
For reference, that error message is `...invalid input syntax for type uuid: "null")...` (HTTP 500).

Reply:
> > Can you expand upon what you mean here? Do you mean the error should be reported differently from PgUtil?
> 
> I was thinking more along the lines of changing the tests to look specifically for their table:value (aka: leave PgUtil alone).

I'm ok with the tests changing to describe what the new error message is. We might not be able to retain the previous message.

> Now that I am re-looking at the error message `value already exists in table loan: 335b7118-76e7-4d0a-9453-3dda29dfe529` is not that much more helpful.

Is this a PostgreSQL error that is forwarded on by PgUtil or is PgUtil translating these error messages?

> The error message `...value already exists in table loan:...` (HTTP 422) does not contain the unique index, and so there is no way to truly infer this.
> 
> The error message `...duplicate key value violates unique constraint "loan_itemid_idx_unique"...` (HTTP 400) does contain the index and can be inferred.

Ok, so there are some cases where the error contains enough to figure out the specific cause?

Reply:
> Is this a PostgreSQL error that is forwarded on by PgUtil or is PgUtil translating these error messages?

This is a message _created by PgUtil_, see the link in my earlier comments (`PgUtil.java#L332`). The `responseUniqueViolation` method. Now that I am glancing at that method, it makes me wonder if that code is somehow failing. There is a regex match being used and if that regex fails to match, then it returns the custom generated response with either a 422 or a non-422 status.

> Ok, so there are some cases where the error contains enough to figure out the specific cause?

Yes and considered my above comment in regards to `responseUniqueViolation`, I am wondering if this happens in cases where the regex matches as intended.

Reply:
> > Is this a PostgreSQL error that is forwarded on by PgUtil or is PgUtil translating these error messages?
> 
> This is a message _created by PgUtil_, see the link in my earlier comments (`PgUtil.java#L332`). The `responseUniqueViolation` method. Now that I am glancing at that method, it makes me wonder if that code is somehow failing. There is a regex match being used and if that regex fails to match, then it returns the custom generated response with either a 422 or a non-422 status.

If you believe something is failing for the wrong reasons that should be investigated.

If it is likely to be only a change in message, then I agree, we should just change the expected message in the tests so they accurately describe what is happening.